File: 33.Search_in_Rotated_Sorted_Array.cpp
class Solution {
    public:
    int search(vector<int>& nums, int target) {
        if (nums.empty()) return -1;
        int start = 0;
        int end = nums.size()-1;
        int mid;
        while (start+1 < end){
            mid = (start+end)/2;
            if(target == nums[mid]) return mid;
            if(nums[start] < nums[mid]){
                if (target>=nums[start] && target<nums[mid]) end = mid;
                else start = mid;
            }
            else {
                if(target>nums[mid] && target<=nums[end]) start = mid;
                else end = mid;
            }
        }
        if(nums[start] == target) return start;
        else if (nums[end] == target) return end;
        else return -1;
    }
};

File: 33.SearchinRotatedSortedArray.py
'''
先判断哪半边是sorted order
if target 在sorted 半边，更新 left/right
else 在另外半边
'''


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        start, end = 0, len(nums)-1
        while start < end:
            mid = (start + end) // 2
            if nums[mid] == target:
                return mid
            elif nums[start] < nums[mid]:
                if target < nums[mid] and target >= nums[start]:
                    end = mid
                else:
                    start = mid+1
            elif nums[start] > nums[mid]:
                if target > nums[mid] and target <= nums[end]:
                    start = mid+1
                else:
                    end = mid
            else:
                start += 1
        return start if nums[start] == target else -1
